src/tox_travis/toxenv.py

```diff
@@ -10,43 +10,76 @@
 from .utils import TRAVIS_FACTORS, parse_dict
 
 
-def default_toxenv():
-    """Default TOXENV automatically based on the Travis environment."""
-    if 'TOXENV' in os.environ:
+def default_toxenv(config):
+    """Default envlist automatically based on the Travis environment."""
+    if 'TOXENV' in os.environ or config.option.env:
         return  # Skip any processing if already set
 
-    config = py.iniconfig.IniConfig('tox.ini')
+    ini = py.iniconfig.IniConfig('tox.ini')
 
     # Find the envs that tox knows about
-    declared_envs = get_declared_envs(config)
+    declared_envs = get_declared_envs(ini)
 
     # Find all the envs for all the desired factors given
-    desired_factors = get_desired_factors(config)
+    desired_factors = get_desired_factors(ini)
 
     # Reduce desired factors
     desired_envs = ['-'.join(env) for env in product(*desired_factors)]
 
     # Find matching envs
     matched = match_envs(declared_envs, desired_envs,
                          passthru=len(desired_factors) == 1)
-    os.environ.setdefault('TOXENV', ','.join(matched))
 
+    # Make the envconfig for undeclared matched envs
+    autogen_envconfigs(config, set(matched) - set(config.envconfigs))
```

Comment:
This part puzzles me, comparing this to #61, I don't understand what exactly `autgen_envconfigs` does and why it's necessary. It's not clear to me what's depending on the generated envconfigs, and why this wasn't necessary in #61.

Reply:
This is a case that #61 didn't cover, and exists for backward compatibility with the first release of tox-travis. With that first release, you would declare what environments got run for each version of Python. It didn't look at the environments that were listed at all, it just sent the string in the configuration verbatim, through the TOXENV environment variable. If the specified environment is not declared in the tox config, Tox will still happily run that environment, doing the best it can based on the settings it has.

When we switched to matching based on factors, so that by default on Python 3.4 all envs with `py34` in the name would run, that obviously was inherently in conflict with the original mechanism. However, I was able to maintain partial backward compatibility by saying that if there was only one "factor" being matched (`python`, which is automatic), and no environments matched it, then whatever the setting is will be used verbatim, in the same way as was done in the original version.

This is a wart, but fixing it means a breaking change. This workaround allows me to get this change in without having to worry about a breaking change at this time.

Reply:
Oh, and as far as _what_ it actually does. Before the tox_configure hook, Tox looks at all the environments, and creates a envconfig for each one. The envconfig is the configuration of how to run the environment, with default settings resolved. What this does is notice when we've ended up with an envconfig that Tox didn't know about to begin with, and make an envconfig for it the same way Tox would have if it had been specified in `TOXENV`.

Reply:
> This is a wart, but fixing it means a breaking change. This workaround allows me to get this change in without having to worry about a breaking change at this time.

Gotcha, this is compatibility code, and would presumably not be present in the next major release.